File: June-LeetCoding-Challenge/12-Insert-Delete-GetRandom/Insert-Delete-GetRandom.cpp
class RandomizedSet {
public:
    unordered_map<int,int> m;
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        return (++m[val])==1;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(m.find(val)==m.end()) return false;
        m.erase(val);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
       return next(begin(m), rand()%m.size())->first;
    }
};

class RandomizedSet {
    unordered_map<int, int> hm;
    vector<int> datas;
public:
    /** Initialize your data structure here. */
    RandomizedSet() { }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (hm.find(val) != hm.end())
            return false;
        hm[val] = datas.size();
        datas.push_back(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (hm.find(val) == hm.end())
            return false;
        hm[datas.back()] = hm[val];
        datas[hm[val]] = datas.back();
        hm.erase(val);
        datas.pop_back();
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return datas[random()%datas.size()];
    }
};

File: June-LeetCoding-Challenge/12-Insert-Delete-GetRandom/Insert-Delete-GetRandom.kt
import kotlin.random.Random

class InsertDeleteGetRandomO1Kotlin380 {
    /** Initialize your data structure here. */
    private val arrayList: MutableList<Int> = ArrayList()
    private val map: MutableMap<Int, Int> = HashMap()

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    fun insert(`val`: Int): Boolean {
        if (map.containsKey(`val`)) {
            return false
        }
        map[`val`] = arrayList.size
        arrayList.add(`val`)
        return true
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    fun remove(`val`: Int): Boolean {
        if (!map.containsKey(`val`)) {
            return false
        }
        val removeIndex = map.getValue(`val`)
        if (removeIndex != arrayList.size - 1) {
            arrayList[removeIndex] = arrayList[arrayList.size - 1]
            map[arrayList[arrayList.size - 1]] = removeIndex
        }
        arrayList.removeAt(arrayList.size - 1)
        map.remove(`val`)
        return true
    }

    /** Get a random element from the set. */
    fun getRandom(): Int {
        return arrayList[Random.Default.nextInt(arrayList.size)]
    }
}

fun main() {
    val solution = InsertDeleteGetRandomO1Kotlin380()
    println(solution.insert(1))
    println(solution.remove(2))
    println(solution.insert(2))
    println(solution.getRandom())
    println(solution.remove(1))
    println(solution.insert(2))
}